src/ocl/batchnormocl.cpp

```diff
@@ -262,23 +262,63 @@ void BatchNormForwardTraining(Handle& handle,
                 std::string program_name;
                 std::string parms;
 
+                unsigned int rd_blk = 1;
+                unsigned int grprd;
+                unsigned int mio_bn_REM4;
+                unsigned int mio_bn_LESS4;
+                unsigned int mio_bn_CHUNK4;
+                unsigned int mio_bn_REMOUT4;
+                unsigned int mio_bn_LESSOUT4;
+                unsigned int mio_bn_REM;
+                unsigned int mio_bn_LESS;
+                unsigned int mio_bn_CHUNK;
+                unsigned int mio_bn_REMOUT;
+                unsigned int mio_bn_LESSOUT;
+                unsigned int mio_max_read = 2;
+                if(in_cstride >= 4096)
+                {
+                    mio_max_read = 3;
+                }
+                grprd           = (xlocalsize * rd_blk * 4);
+                mio_bn_REM4     = (in_nhw - ((in_nhw / grprd) * grprd));
+                mio_bn_LESS4    = (in_nhw - mio_bn_REM4);
+                mio_bn_CHUNK4   = (mio_max_read * grprd);
+                mio_bn_REMOUT4  = (in_nhw - ((in_nhw / mio_bn_CHUNK4) * mio_bn_CHUNK4));
+                mio_bn_LESSOUT4 = (in_nhw - mio_bn_REMOUT4);
+                mio_bn_REM      = (in_nhw - ((in_nhw / xlocalsize) * xlocalsize));
+                mio_bn_LESS     = (in_nhw - mio_bn_REM);
+                mio_bn_CHUNK    = (mio_max_read * xlocalsize);
+                mio_bn_REMOUT   = (in_nhw - ((in_nhw / mio_bn_CHUNK) * mio_bn_CHUNK));
+                mio_bn_LESSOUT  = (in_nhw - mio_bn_REMOUT);
+
                 kernel_name  = "MIOpenBatchNormFwdTrainSpatial";
                 program_name = "MIOpenBatchNormFwdTrainSpatial.cl";
 
-                parms = " -DMIOPEN_USE_FP16=" + std::to_string(static_cast<int>(bfp16parm)) +
-                        " -DMIOPEN_USE_FP32=" + std::to_string(static_cast<int>(bfp32parm)) +
-                        " -DMIOPEN_USE_FPMIX=" + std::to_string(static_cast<int>(bfpmixparm)) +
-                        " -DMIO_SAVE_MEAN_VARIANCE=" +
-                        std::to_string(static_cast<int>(resultsave)) + " -DMIO_RUNNING_RESULT=" +
-                        std::to_string(static_cast<int>(resultrunning)) + " -DMIO_BN_N=" +
-                        std::to_string(n) + " -DMIO_BN_C=" + std::to_string(c) + " -DMIO_BN_HW=" +
-                        std::to_string(in_cstride) + " -DMIO_BN_NHW=" + std::to_string(in_nhw) +
-                        " -DMIO_BN_CHW=" + std::to_string(in_nstride) + " -DMIO_BN_NCHW=" +
-                        std::to_string(in_nchw) + " -DMIO_BN_LDS_SIZE=" + std::to_string(ldsnogcn) +
-                        " -DMIO_BN_LDSGCN_SIZE=" + std::to_string(ldsgcn) + " -DMIO_BN_VARIANT=" +
-                        std::to_string(variant) + " -DMIO_BN_GRP0=" + std::to_string(xlocalsize) +
-                        " -DMIO_BN_GRP1=" + std::to_string(ylocalsize) + " -DMIO_BN_GRP2=" +
-                        std::to_string(zlocalsize);
+                parms =
```

Comment:
This does not fulfill the requirement of reducing compiles. If you pass any of these into the compiler as -D flags it will trigger a new compile.
e.g., `mio_max_read` should be a parameter into the argument list.